### PR TITLE
Themes dark mode example

### PR DESCRIPTION
**Pull request by vie10**
I didn't found any examples how to use MUI themes in Kotlin, so I think it will be useful

### CODE REVIEW

**Comment by vie10** [comment]
Screenshots after commit:

File: src/main/kotlin/example/SidebarOpenedModule.kt [hunk 1]
```diff
@@ -1,21 +1,29 @@
 package example
 
+import mui.system.ThemeProvider
 import react.*
 
 data class SidebarOpened(
     val value: Boolean,
     val setValue: StateSetter<Boolean>,
+    val darkMode: Boolean,
```
Comment (aerialist7) [inline src/main/kotlin/example/SidebarOpenedModule.kt:9]: Single Responsibility Principle is violated here.
Theme has no relation to `SidebarOpenedContext`.

File: src/main/kotlin/example/SidebarOpenedModule.kt [hunk 2]
```diff
@@ -1,21 +1,29 @@
...
-    val sidebarOpened = useMemo(open, setOpen) {
-        SidebarOpened(open, setOpen)
+    val sidebarOpened = useMemo(open, setOpen, darkMode, setDarkMode) {
+        SidebarOpened(open, setOpen, darkMode, setDarkMode)
     }
 
     SidebarOpenedContext.Provider(sidebarOpened) {
-        props.children()
+        val newTheme = if (darkMode) themeDark else themeLight
+        ThemeProvider {
```
Comment (aerialist7) [inline src/main/kotlin/example/SidebarOpenedModule.kt:24]: `ThemeProvider` shouldn't be used in `SidebarOpenedModule` component.
Please check the examples from https://mui.com/customization/theming/#api

File: src/main/kotlin/example/Themes.kt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package example
+
+import mui.material.styles.ThemeOptions
+import mui.material.styles.createTheme
+
+val themeDark = createTheme(js("({palette: { mode: 'dark' }})").unsafeCast<ThemeOptions>())
```
Comment (aerialist7) [inline src/main/kotlin/example/Themes.kt:6]: `themeDark` -> `darkTheme`?

File: src/main/kotlin/example/Themes.kt [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+val themeLight = createTheme(js("({palette: { mode: 'light' }})").unsafeCast<ThemeOptions>())
```
Comment (aerialist7) [inline src/main/kotlin/example/Themes.kt:7]: `themeLight` -> `lightTheme`?

File: src/main/kotlin/example/Themes.kt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package example
+
+import mui.material.styles.ThemeOptions
+import mui.material.styles.createTheme
+
```
Comment (aerialist7) [inline src/main/kotlin/example/Themes.kt:5]: Please do not use pure `js` function.
There are better options.

File: src/main/kotlin/example/Header.kt [hunk 6]
```diff
@@ -57,6 +59,32 @@ val Header = FC<Props> {
...
+            Tooltip {
+                title = ReactNode(
+                    if (darkMode) {
+                        "Switch to light mode"
+                    } else {
+                        "Switch to dark mode"
+                    }
+                )
+
+                IconButton {
```
Comment (aerialist7) [inline src/main/kotlin/example/Header.kt:71]: Try to use `Switch` component. 
Seems to be the most acceptable solution for theming.

File: src/main/kotlin/example/Header.kt [hunk 7]
```diff
@@ -20,6 +19,9 @@ import react.router.useLocation
 import react.useContext
 
 val Header = FC<Props> {
+    val sidebarOpenedData = useContext(SidebarOpenedContext)
+    val darkMode = sidebarOpenedData.darkMode
```
Comment (aerialist7) [inline src/main/kotlin/example/Header.kt:23]: Destructure it, please.

File: src/main/kotlin/example/Header.kt [hunk 8]
```diff
@@ -20,6 +19,9 @@ import react.router.useLocation
 import react.useContext
 
 val Header = FC<Props> {
+    val sidebarOpenedData = useContext(SidebarOpenedContext)
```
Comment (aerialist7) [inline src/main/kotlin/example/Header.kt:22]: Copypaste detected.

File: src/main/kotlin/example/SidebarOpenedModule.kt [hunk 9]
```diff
@@ -1,21 +1,29 @@
...
     val (open, setOpen) = useState(true)
-    val sidebarOpened = useMemo(open, setOpen) {
-        SidebarOpened(open, setOpen)
+    val sidebarOpened = useMemo(open, setOpen, darkMode, setDarkMode) {
+        SidebarOpened(open, setOpen, darkMode, setDarkMode)
     }
 
     SidebarOpenedContext.Provider(sidebarOpened) {
-        props.children()
+        val newTheme = if (darkMode) themeDark else themeLight
```
Comment (aerialist7) [inline src/main/kotlin/example/SidebarOpenedModule.kt:23]: Inline it, please.

**Review by aerialist7 — CHANGES_REQUESTED** [review]
Thank you for your PR! This is very useful. 
But please fix the remarks on the review.

**Comment by vie10** [comment]
Thx for Wonderful tips! i got a lot of new knowledges with your help (:

File: README.md [hunk 10]
```diff
@@ -1,8 +1,9 @@
 # [Kotlin MUI Showcase](https://karakum-team.github.io/kotlin-mui-showcase/)
 
-> ![Accordion](.doc/accordion.jpg)
-> ![Alert](.doc/alert.jpg)
-> ![Dialog](.doc/dialog.jpg)
-> ![Image List](.doc/image-list.jpg)
-> ![Menu](.doc/menu.jpg)
-> ![Table](.doc/table.jpg)
+> ![Appbar Light](.doc/appbar_light.png)
```
Comment (aerialist7) [inline README.md:3]: Please do not change the screenshots.

File: src/main/kotlin/example/ColorMode.kt [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+            mode = "dark"
+        }
+    )
+}
+
+typealias ColorModeState = StateInstance<PaletteMode>
+
+val ColorModeContext = createContext<ColorModeState>()
+
+val ColorModeModule = FC<PropsWithChildren> { props ->
```
Comment (aerialist7) [inline src/main/kotlin/example/ColorMode.kt:24]: Please create single `ThemeModule` which encapsulates MUI `ThemeProvider` and your `ColorModeContext` (also plese rename it to `ThemeContext`).

File: src/main/kotlin/example/Header.kt [hunk 12]
```diff
@@ -4,22 +4,19 @@ import csstype.*
...
+import react.*
 import react.dom.aria.AriaHasPopup.`false`
 import react.dom.aria.ariaHasPopup
 import react.dom.aria.ariaLabel
 import react.dom.html.ReactHTML
 import react.router.useLocation
-import react.useContext
 
 val Header = FC<Props> {
+    val (paletteMode, setPaletteMode) = useContext(ColorModeContext)
```
Comment (aerialist7) [inline src/main/kotlin/example/Header.kt:19]: And here you'll get a theme from `ThemeContext`
`var theme = useContext(ThemeContext)`

File: src/main/kotlin/example/Header.kt [hunk 13]
```diff
@@ -83,7 +94,7 @@ val Header = FC<Props> {
                         var name = lastPathname
                             .split("-")
                             .asSequence()
-                            .map { it.replaceFirstChar { it.titlecase() } }
+                            .map { it.replaceFirstChar { firstChar -> firstChar.titlecase() } }
```
Comment (aerialist7) [inline src/main/kotlin/example/Header.kt:97]: Seems redundant

File: src/main/kotlin/example/component/Switches.kt [hunk 14]
```diff
@@ -19,5 +22,9 @@ val SwitchesShowcase = FC<Props> {
         Switch {
             disabled = true
         }
+        Switch {
```
Comment (aerialist7) [inline src/main/kotlin/example/component/Switches.kt:25]: Redundant?
If no, please move it to separate PR

File: src/main/kotlin/example/ColorMode.kt [hunk 15]
```diff
@@ -0,0 +1,45 @@
+package example
+
+import kotlinext.js.jso
+import mui.material.styles.*
+import react.*
+
+enum class PaletteMode(val paletteOptions: PaletteOptions) {
```
Comment (aerialist7) [inline src/main/kotlin/example/ColorMode.kt:7]: Check file `Sizes.kt` and reorganize it in the same way.
Seems it can be called `Themes`

**Review by aerialist7 — CHANGES_REQUESTED** [review]
On the right track, thank you!
Found some minor remarks.

File: src/main/kotlin/example/ColorMode.kt (hunk 11 above)
Comment (vie10) [reply src/main/kotlin/example/ColorMode.kt:24]: You mean something like it? Or an another way to do the double encapsulation?

```kotlin
val ThemeModule = FC<PropsWithChildren> { props ->
    val theme = useState(Themes.Light)

    ThemeContext.Provider(theme) {
        ThemeProvider {
            this.theme = theme
            props.children()
        }
    }
}
```

Comment (aerialist7) [reply src/main/kotlin/example/ColorMode.kt:24]: Exactly this

File: src/main/kotlin/example/Header.kt [hunk 16]
```diff
@@ -4,22 +4,22 @@ import csstype.*
...
+import react.*
 import react.dom.aria.AriaHasPopup.`false`
 import react.dom.aria.ariaHasPopup
 import react.dom.aria.ariaLabel
 import react.dom.html.ReactHTML
 import react.router.useLocation
-import react.useContext
 
 val Header = FC<Props> {
+    val (theme, setTheme) = useContext(ThemeContext)
```
Comment (aerialist7) [inline src/main/kotlin/example/Header.kt:22]: I found cool hook `useTheme` - https://mui.com/styles/advanced/#accessing-the-theme-in-a-component
Seems that we can use it instead of our `ThemeModule` component :)

**Review by aerialist7 — CHANGES_REQUESTED** [review]
Silver bullet detected - https://mui.com/styles/advanced/#accessing-the-theme-in-a-component 😀

File: src/main/kotlin/example/Header.kt (hunk 16 above)
Comment (vie10) [reply src/main/kotlin/example/Header.kt:22]: Looks like there no way to set theme without the module, I see just way to access context theme by useTheme()
Do you've idea how to do it?

Comment (aerialist7) [reply src/main/kotlin/example/Header.kt:22]: Yep, seems we really can't

**Review by aerialist7 — APPROVED** [review]

**Review by aerialist7 — APPROVED** [review]

**Comment by aerialist7** [comment]
@vie10, thank you! 
You are welcome with future PRs 😉

For example you can help us with adaptive stylization (for mobile devices) or you can add other components to the showcase

**Comment by vie10** [comment]
> @vie10, thank you! You are welcome with future PRs 😉
> 
> For example you can help us with adaptive stylization (for mobile devices) or you can add other components to the showcase

I'm glad to hear that! 😆
Thx for new experience :)
